### PR TITLE
adding vg api

### PR DESCRIPTION
**Pull request by matancarmeli7**

### CODE REVIEW

**Review by nixpanic — COMMENTED** [review]
There could be many more comments in the go-code. A description of the PR is missing, and would be nice to have references to the CSI/Kubernetes proposals.

File: go.mod [hunk 1]
```diff
@@ -82,4 +82,4 @@ require (
 	sigs.k8s.io/yaml v1.3.0 // indirect
 )
 
-replace github.com/csi-addons/spec => github.com/matancarmeli7/spec v0.0.0-20230703063652-bd35954246e7
+replace github.com/csi-addons/spec => github.com/matancarmeli7/spec v0.0.0-20230711062057-d8d78175c552
```
[the remark this answers is not in the document]
Comment (matancarmeli7) [reply go.mod:85]: agreed, I will change it once the PR in the spec will be changed

File: controllers/volumegroup.storage/pkg/config/config.go [hunk 2]
```diff
@@ -43,16 +45,19 @@ func GetDriverConfig(driverName string, connPoll *conn.ConnectionPool) (*DriverC
 				continue
 			}
 
-			if cap.GetVolumeGroup().GetType() == identity.Capability_VolumeGroup_LIMIT_VOLUME_TO_ONE_VG {
+			if cap.GetVolumeGroup().GetType() == identity.Capability_VolumeGroup_LIMIT_VOLUME_TO_ONE_VOLUME_GROUP {
```
[the remark this answers is not in the document]
Comment (matancarmeli7) [reply controllers/volumegroup.storage/pkg/config/config.go]: done

File: controllers/volumegroup.storage/pkg/config/config.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	conn "github.com/csi-addons/kubernetes-csi-addons/internal/connection"
+	"github.com/csi-addons/spec/lib/go/identity"
+)
+
+type DriverConfig struct {
+	DriverEndpoint       string
+	DriverName           string
+	MultipleVGsToPVC     string
+	DeletePVCsOnVGDelete string
+	ModifyVolumeGroup    string
```
[the remark this answers is not in the document]
Comment (matancarmeli7) [reply controllers/volumegroup.storage/pkg/config/config.go]: done

**Comment by nixpanic** [comment]
@mergifyio rebase

File: controllers/volumegroup.storage/pkg/config/config.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	// provided in the VolumeReplication spec; so that corresponding
+	// operations can be performed.
+	for _, v := range conns {
+		for _, cap := range v.Capabilities {
+			// validate if VOLUME_REPLICATION capability is supported by the driver.
+			if cap.GetVolumeGroup() == nil {
+				continue
+			}
+
+			switch capType := cap.GetVolumeGroup().GetType(); capType {
```
Comment (nixpanic) [inline controllers/volumegroup.storage/pkg/config/config.go:48]: maybe write this like
```go
switch cap.GetVolumeGroup().GetType() {
```
There is no use for `capType` as additional variable.

File: go.mod (hunk 1 above)
[the remark this answers is not in the document]
Comment (nixpanic) [reply go.mod:85]: this can be corrected now

**Review by nixpanic — COMMENTED** [review]
There most likely are functions under controllers/volumegroup.storage/utils that are available somewhere else, or would be suitable for other controllers too. This should probably be cleaned up at a later time.

At the moment I am missing how the communication to the CSI-driver is done. The kubernetes-csi-addons operator communicates over gRPC to the csi-addons-sidecar. The sidecar can be running alongside the CSI-controller (provisioner) or the CSI-nodeplugin. VolumeGroups most likely need to only call function of the CSI-controller. The protocol between the operator and sidecar should be rather simple as the sidecar can resolve the Kubernetes objects (PVC, PV, Secrets, ...). There is no `.proto` file added for this communication under https://github.com/csi-addons/kubernetes-csi-addons/tree/main/internal/proto , which is what I would expect.

**Comment by nixpanic** [comment]
This is too complex to include, #588 is a simpler API that lets CSI-drivers handle the grouping of volumes how they like.

**Pull request closed without merging**